### PR TITLE
HPCC-22826 Crypto helper wants Digital Signatures configured

### PR DESCRIPTION
**Pull request by RussWhitehead**
Incorrect logic checking for existence of public key file spec

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@ghalliday Please review

File: plugins/cryptolib/cryptolib.cpp [hunk 1]
```diff
@@ -437,7 +437,7 @@ class CDSMCache
         else
         {
             IDigitalSignatureManager * pDSM = nullptr;
-            if (!isEmptyString(pubKeyFS) && !isEmptyString(privKeyFS))
+            if (!isEmptyString(pubKeyFS) || !isEmptyString(privKeyFS))
```
Comment (ghalliday) [inline plugins/cryptolib/cryptolib.cpp]: @RussWhitehead  A reasonable change, but why does this fix the issue that was reported?  In that case they were both provided.

**Comment by RussWhitehead** [comment]
@ghalliday Because when the workunit code calls "clPKISign", it only provides the private key file. Therefore, clPKISign searches for the Digital Signature Manager as follows, passing NULL for the unknown and unneeded Public Key File

CRYPTOLIB_API void CRYPTOLIB_CALL clPKISign(size32_t & __lenResult,void * & __result,
                                        const char * pkalgorithm,
                                        const char * privatekeyfile,
                                        const char * passphrase,
                                        size32_t lenInputdata,const void * inputdata)
{
    verifyPKIAlgorithm(pkalgorithm);//TODO extend cryptohelper to support more algorithms
    Owned<IDigitalSignatureManager> pDSM = g_DSMCache.getInstance(pkalgorithm, nullptr, nullptr, privatekeyfile, nullptr, passphrase);

**Comment by RussWhitehead** [comment]
@ghalliday Above I should have mentioned, the library call to sign and verify only take as an argument the key they need

    EXPORT DATA Sign( DATA inputData) := FUNCTION
        RETURN lib_cryptolib.CryptoLib.Sign( pkAlgorithm, privateKeyFile, passphrase, inputData);
    END;

    EXPORT BOOLEAN VerifySignature(DATA signature, DATA signedData) := FUNCTION
        RETURN lib_cryptolib.CryptoLib.VerifySignature( pkAlgorithm, publicKeyFile, passphrase, signature, signedData);
    END;

**Comment by ghalliday** [comment]
@RussWhitehead 
ok, thanks. 

That suggests that the IDigitalSignatureManager interface would be better split into two, but something to think about separately.

**Comment by ghalliday** [comment]
Please can you rebase onto 7.6.0, since it may as well be included in that version.

**Comment by RussWhitehead** [comment]
@ghalliday Rebased to 7.6.0, please merge

**Comment by richardkchapman** [comment]
@RussWhitehead Can you double check your rebase - seems to have some other commits in too

**Comment by AttilaVamos** [comment]
@RussWhitehead I am sure the schedule2 failure on Roxie is not related to this PR.

**Pull request closed without merging**